tests/timer_ut/timer_ut.c

```diff
@@ -0,0 +1,229 @@
+    ASSERT_IS_NOT_NULL(test_serialize_mutex);
+
+    result = umock_c_init(on_umock_c_error);
+    ASSERT_ARE_EQUAL(int, 0, result, "umock_c_init");
+
+    result = umocktypes_stdint_register_types();
+    ASSERT_ARE_EQUAL(int, 0, result, "umocktypes_stdint_register_types failed");
+
+    result = umocktypes_charptr_register_types();
+    ASSERT_ARE_EQUAL(int, 0, result, "umocktypes_charptr_register_types failed");
+
+    result = umocktypes_bool_register_types();
+    ASSERT_ARE_EQUAL(int, 0, result, "umocktypes_bool_register_types failed");
+
+    REGISTER_GLOBAL_MOCK_HOOK(gballoc_malloc, my_gballoc_malloc);
+    REGISTER_GLOBAL_MOCK_HOOK(gballoc_free, my_gballoc_free);
+}
+
+TEST_SUITE_CLEANUP(suite_cleanup)
+{
+    umock_c_deinit();
+    umock_c_negative_tests_deinit();
+    TEST_MUTEX_DESTROY(test_serialize_mutex);
+}
+
+TEST_FUNCTION_INITIALIZE(init)
+{
+    if (TEST_MUTEX_ACQUIRE(test_serialize_mutex))
+    {
+        ASSERT_FAIL("Could not acquire test serialization mutex.");
+    }
+
+    umock_c_reset_all_calls();
+}
+
+/* timer_create */
+TEST_FUNCTION(timer_create_malloc_fails)
+{
+    //arrange
+    STRICT_EXPECTED_CALL(malloc(IGNORED_NUM_ARG))
+        .SetReturn(NULL);
+
+    //act
+    TIMER_HANDLE timer = timer_create();
+
+    //assert
+    ASSERT_IS_NULL(timer, "timer_create failed.");
+}
+
+static void test_timer_create_success_expectations(void)
+{
+    LARGE_INTEGER frequency, start_time;
+    frequency.QuadPart = 10;
+    start_time.QuadPart = 1;
+    STRICT_EXPECTED_CALL(malloc(IGNORED_NUM_ARG));
+    STRICT_EXPECTED_CALL(mocked_QueryPerformanceFrequency(IGNORED_PTR_ARG))
+        .CopyOutArgumentBuffer_lpFrequency(&frequency, sizeof(frequency));
+    STRICT_EXPECTED_CALL(mocked_QueryPerformanceCounter(IGNORED_PTR_ARG))
+        .CopyOutArgumentBuffer_lpPerformanceCount(&start_time, sizeof(start_time));
+}
+
+TEST_FUNCTION(timer_create_succeeds)
+{
+    //arrange
+    test_timer_create_success_expectations();
+
+    //act
+    TIMER_HANDLE timer = timer_create();
+
+    //assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_IS_NOT_NULL(timer, "timer_create failed.");
+
+    //cleanup
+    timer_destroy(timer);
+}
+
+TEST_FUNCTION(timer_get_elapsed_fails_if_timer_is_null)
+{
+    //arrange
+    //act
+    double start = timer_get_elapsed(NULL);
+
+    //assert
+    ASSERT_ARE_EQUAL(double, -1, start);
+}
+
+TEST_FUNCTION(timer_get_elapsed_success)
+{
+    //arrange
+    LARGE_INTEGER stop_time;
+    stop_time.QuadPart = 100;
+    test_timer_create_success_expectations();
+    TIMER_HANDLE timer = timer_create();
+    umock_c_reset_all_calls();
+    STRICT_EXPECTED_CALL(mocked_QueryPerformanceCounter(IGNORED_PTR_ARG))
+        .CopyOutArgumentBuffer_lpPerformanceCount(&stop_time, sizeof(stop_time));
+    
+    //act
+    double elapsed_time = timer_get_elapsed(timer);
+
+    //assert
+    ASSERT_ARE_EQUAL(double, 9.9, elapsed_time);
```

Comment:
> ASSERT_ARE_EQUAL(double, 9.9, elapsed_time); [](start = 3, length = 45)

Something tells me this may require a fix for the `double_Compare` defined by ctest.

int double_Compare(double left, double right)
{
    return left != right;
}


Because (9.9 == 99/10) is not necessarily true.

`double_Compare` may need to allow some small epsilon difference to make this behave as expected. Or maybe this test should just do that (e.g. make sure `elapsed_time` is > (9.9-epsilon) and `elapsed_time` is < (9.9+epsilon)).

Reply:
Also, ctestrunnerswitcher doesn't seem to do the right thing for floating point compare when CppUnit is used.

```
#define ASSERT_ARE_EQUAL(type, A, B, ...) \
    do \
    { \
        char* ctrs_message = CONSTRUCT_CTRS_MESSAGE(__VA_ARGS__); \
        std::wstring cppUnitTestMessage = ToString(ctrs_message); \
        ctrs_sprintf_free(ctrs_message); \
        Assert::AreEqual((type)(A), (type)(B), cppUnitTestMessage.c_str()); \
    } while ((void)0, 0)
```

But the signature has a `tolerance` that we don't pass in.

```
  // double AreNotEqual comparison:
  static void AreNotEqual(double notExpected, double actual, double tolerance, const wchar_t* message = NULL, const __LineInfo* pLineInfo = NULL)
  {
   double diff = notExpected - actual;
   FailOnCondition(fabs(diff) > fabs(tolerance), NOT_EQUALS_MESSAGE(notExpected, actual, message), pLineInfo);
  }
```


---
In reply to: [317311507](https://github.com/Azure/azure-c-shared-utility/pull/362#discussion_r317311507) [](ancestors = 317311507)

Reply:
had a discussion with Matt, keeping it as is to keep the test simple.